### PR TITLE
Added notification module

### PR DESCRIPTION
**Pull request by adhyans**
This module will allow developers to integrate notifications with their functionality. Here are some wikis on which you can read more about this module.

1) [Notification Module Design](https://mifosforge.jira.com/wiki/pages/viewpage.action?pageId=133070890)
2) [Notification API Developer User Guide](https://mifosforge.jira.com/wiki/display/MDZ/Notification+API+Developer+User+Guide)
3) [Configure Notifications-End User guide](https://mifosforge.jira.com/wiki/display/docs/Configure+Notifications)

### CODE REVIEW

File: fineract-provider/gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,6 +1,6 @@
-#Fri Jan 29 13:04:36 IST 2016
+#Tue Jul 19 20:53:11 IST 2016
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.10-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-2.10-all.zip
```
Comment (nazeer1100126) [inline fineract-provider/gradle/wrapper/gradle-wrapper.properties:6]: why do we need this change?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/SearchParameters.java [hunk 2]
```diff
@@ -75,6 +75,15 @@ public static SearchParameters forClients(final String sqlSearch, final Long off
                 orderBy, sortOrder, staffId, accountNo, loanId, savingsId, orphansOnly, isSelfUser);
     }
 
+    public static SearchParameters forNotifications(final Integer offset, final Integer limit,
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/SearchParameters.java:78]: This new method is not required. You can use SearchParameters.forPagination(...) method

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/filter/TenantAwareBasicAuthenticationFilter.java [hunk 3]
```diff
@@ -167,6 +168,12 @@ protected void onSuccessfulAuthentication(HttpServletRequest request,
     		throws IOException {
     	super.onSuccessfulAuthentication(request, response, authResult);
 		AppUser user = (AppUser) authResult.getPrincipal();
+
+        if (notificationReadPlatformService.hasUnreadNotifications(user.getId())) {
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/filter/TenantAwareBasicAuthenticationFilter.java:172]: Is notification module supported only for basicauth profile? Don't we support for oauth profile?

File: fineract-provider/src/main/java/org/apache/fineract/notification/api/NotificationApiResource.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import javax.ws.rs.*;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.UriInfo;
+
+@Path("/notifications")
+@Component
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/notification/api/NotificationApiResource.java:21]: scope should be singleton

File: fineract-provider/src/main/resources/sql/migrations/core_db/V254__General_Ledger_Report.sql [hunk 5]
```diff
@@ -39,5 +39,5 @@ INSERT INTO `stretchy_report_parameter` (`report_id`, `parameter_id`, `report_pa
    INSERT INTO `stretchy_report_parameter` (`report_id`, `parameter_id`, `report_parameter_name`) VALUES
  ((select sr.id from stretchy_report sr where sr.report_name='GeneralLedgerReport'), 
  (select sp.id from stretchy_parameter sp where sp.parameter_name='OfficeIdSelectOne'), 
-  'office');
+  'officeId');
```
Comment (nazeer1100126) [inline fineract-provider/src/main/resources/sql/migrations/core_db/V254__General_Ledger_Report.sql:42]: You should not modify the existing migration files

File: fineract-provider/src/main/resources/sql/migrations/core_db/V234__opening_balaces_setup.sql [hunk 6]
```diff
@@ -36,7 +36,7 @@ INSERT INTO `c_configuration`
 	)
 	VALUES 
 	(
-		'office-opening-balances-contra-account', 
+		'officeId-opening-balances-contra-account',
```
Comment (nazeer1100126) [inline fineract-provider/src/main/resources/sql/migrations/core_db/V234__opening_balaces_setup.sql:39]: don't modify existing migration files

File: fineract-provider/src/main/resources/sql/migrations/core_db/V228__entity_to_entity_access.sql [hunk 7]
```diff
@@ -56,9 +56,9 @@ CREATE TABLE `m_entity_to_entity_access` (
```
Comment (nazeer1100126) [inline fineract-provider/src/main/resources/sql/migrations/core_db/V228__entity_to_entity_access.sql:56]: Don't modify existing migration files

File: fineract-provider/src/main/resources/sql/migrations/core_db/V1__mifosplatform-core-ddl-latest.sql [hunk 8]
```diff
@@ -204,7 +204,7 @@ CREATE TABLE IF NOT EXISTS `m_calendar_instance` (
 
 -- ================= end of reference/lookup tables =============
```
Comment (nazeer1100126) [inline fineract-provider/src/main/resources/sql/migrations/core_db/V1__mifosplatform-core-ddl-latest.sql:205]: Don't modify existing migration files

File: fineract-provider/src/main/resources/sql/migrations/core_db/V108__client_has_transfer_office.sql [hunk 9]
```diff
@@ -17,12 +17,12 @@
 -- under the License.
 --
```
Comment (nazeer1100126) [inline fineract-provider/src/main/resources/sql/migrations/core_db/V108__client_has_transfer_office.sql:19]: Don't modify existing migration files

File: fineract-provider/src/main/java/org/apache/fineract/useradministration/service/PermissionReadPlatformServiceImpl.java [hunk 10]
```diff
@@ -39,11 +46,14 @@
 
     private final JdbcTemplate jdbcTemplate;
     private final PlatformSecurityContext context;
+    private final AppUserRepository appUserRepository;
 
     @Autowired
-    public PermissionReadPlatformServiceImpl(final PlatformSecurityContext context, final RoutingDataSource dataSource) {
+    public PermissionReadPlatformServiceImpl(final PlatformSecurityContext context, final RoutingDataSource dataSource,
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/useradministration/service/PermissionReadPlatformServiceImpl.java:52]: In read platforms don't use jpa repository. Use jdbc template

File: fineract-provider/src/main/java/org/apache/fineract/useradministration/service/AppUserReadPlatformServiceImpl.java [hunk 11]
```diff
@@ -95,10 +95,8 @@ public PlatformSecurityContext getContext() {
         final AppUser currentUser = this.context.authenticatedUser();
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/useradministration/service/AppUserReadPlatformServiceImpl.java:95]: This file should not be part of the PR as I don' see any change in this.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/shareproducts/api/ShareDividendApiResource.java [hunk 12]
```diff
@@ -113,6 +119,16 @@ public String createDividendDetail(@PathParam("productId") final Long productId,
         CommandWrapper commandWrapper = new CommandWrapperBuilder().createShareProductDividendPayoutCommand(productId)
                 .withJson(apiRequestBodyAsJson).build();
         final CommandProcessingResult commandProcessingResult = this.commandsSourceWritePlatformService.logCommandSource(commandWrapper);
+
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/shareproducts/api/ShareDividendApiResource.java:122]: Implement Share account business events and publish on event.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/SavingsAccountsApiResource.java [hunk 13]
```diff
@@ -49,7 +49,12 @@
 import org.apache.fineract.infrastructure.core.serialization.DefaultToApiJsonSerializer;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/SavingsAccountsApiResource.java:49]: Listen to the business events and publish the corresponding notification event.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/SavingsAccountTransactionsApiResource.java [hunk 14]
```diff
@@ -41,14 +41,17 @@
 import org.apache.fineract.infrastructure.core.exception.UnrecognizedQueryParamException;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/SavingsAccountTransactionsApiResource.java:41]: Listen to the business event and publish the notification event

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/RecurringDepositAccountsApiResource.java [hunk 15]
```diff
@@ -51,7 +51,10 @@
 import org.apache.fineract.infrastructure.core.serialization.DefaultToApiJsonSerializer;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/RecurringDepositAccountsApiResource.java:51]: Listen to the business event and publish the notification event

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/FixedDepositAccountsApiResource.java [hunk 16]
```diff
@@ -51,7 +51,10 @@
 import org.apache.fineract.infrastructure.core.serialization.DefaultToApiJsonSerializer;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/api/FixedDepositAccountsApiResource.java:51]: Listen to the business event and publish the notification event

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/note/service/NoteWritePlatformServiceJpaRepositoryImpl.java [hunk 17]
```diff
@@ -174,7 +174,7 @@ private CommandProcessingResult createLoanTransactionNote(final JsonCommand comm
     // return new CommandProcessingResultBuilder() //
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/note/service/NoteWritePlatformServiceJpaRepositoryImpl.java:174]: Why getOffice() method is changed to getOfficeId()?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/api/LoanProductsApiResource.java [hunk 18]
```diff
@@ -18,25 +18,6 @@
  */
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/api/LoanProductsApiResource.java:18]: Listen to the business event and publish the notification event

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/rescheduleloan/api/RescheduleLoansApiResource.java [hunk 19]
```diff
@@ -40,14 +40,17 @@
 import org.apache.fineract.infrastructure.core.exception.UnrecognizedQueryParamException;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/rescheduleloan/api/RescheduleLoansApiResource.java:40]: Listen to the business event and publish the notification event

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java [hunk 20]
```diff
@@ -18,29 +18,7 @@
  */
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java:18]: Listen to the business event and publish the notification event

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanTransactionsApiResource.java [hunk 21]
```diff
@@ -45,13 +45,16 @@
 import org.apache.fineract.infrastructure.core.serialization.ApiRequestJsonSerializationSettings;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanTransactionsApiResource.java:45]: Listen to the business event and publish the notification event

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/group/api/GroupsApiResource.java [hunk 22]
```diff
@@ -53,7 +53,9 @@
 import org.apache.fineract.infrastructure.core.serialization.ToApiJsonSerializer;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/group/api/GroupsApiResource.java:53]: Listen to the business event and publish the notification event

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/group/api/CentersApiResource.java [hunk 23]
```diff
@@ -54,7 +54,9 @@
 import org.apache.fineract.infrastructure.core.serialization.ToApiJsonSerializer;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/group/api/CentersApiResource.java:54]: Listen to the business event and publish the notification event

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/api/ClientsApiResource.java [hunk 24]
```diff
@@ -18,25 +18,6 @@
  */
 package org.apache.fineract.portfolio.client.api;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/api/ClientsApiResource.java:19]: Listen to the business event and publish the notification event

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/accounts/api/AccountsApiResource.java [hunk 25]
```diff
@@ -38,14 +38,18 @@
 import org.apache.fineract.infrastructure.core.serialization.ApiRequestJsonSerializationSettings;
```
Comment (nazeer1100126) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/accounts/api/AccountsApiResource.java:38]: Listen to the business event and publish the notification event

**Comment by nazeer1100126** [comment]
Can you please give more details why do we need to use spring-jms?

**Comment by adhyans** [comment]
I have used jms to publish the generated notifications to a queue(implmented via ActiveMq) and then used the jmslistener to subscribe to that queue and listen for any new notification. More specifically, jms with ActiveMq has been used to make the generation and reception of messages to be asynchronous.
And Thanks for the review :smile: I will work on it.

**Comment by nazeer1100126** [comment]
Have a look at how business events are implemented in Fineract. I believe the same way you can do it asynchronously by listening to required events. By this approach you no need to modify all API resources. 
I hope I am clear enough to you :-)

**Comment by adhyans** [comment]
I think I have followed the same design as the implementation of business events.

> You no need to modify all API resources

I have used only the publish method call from the API resources like 
`this.publisher.publishEvent(Event Object)` similar to what the business design uses i.e `this.businessEventNotifierService.notifyBusinessEventToBeExecuted(some parameters)`

Reasons why me and my mentor(Pranjal) decided  to use an external queue :-
1. if we don't use an external queue we can be at a disadvantage of loosing messages if the listener ceases to stop working. 
2. This design follows pub-sub pattern which is nowadays generally used to design notification module.

I think the only difference is that business event design uses a Map to store the list of published events and my implementation uses an external queue.
So, should I refactor like the one suggested by you or proceed with my implementation ?
Thanks.

**Comment by nazeer1100126** [comment]
You need to listen to the business events and you can push notifications in your message queue. API resourced should not have this changes.

**Comment by adhyans** [comment]
Okay, will do and get back to you.Thanks :)

**Comment by adhyans** [comment]
Hi Nazeer,
What do you mean by 
>  API resourced should not have this changes ?

What I have currently refactored is like initiating the business events in the respective API resources and then listening to those business events. Is it good enough ?

**Comment by nazeer1100126** [comment]
Have a look at current  business events implementation. They won't be triggered from api resources.
